dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +232,412 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+        IBackgroundTaskRegistration registration{};
+
+        // Custom version of std::find since we are looking for the name
+        // of the task not the registration.
+        auto first{ std::begin(tasks) };
+        auto last{ std::end(tasks) };
+        while (first != last)
+        {
+            if (first.Current().Value().Name() == backgroundTaskFullName)
             {
-                {
-                    auto lock = s_activatorInfoLock.lock_shared();
-                    THROW_HR_IF(E_INVALIDARG, s_protocolRegistration);
-                }
-
-                // AppId generated by PushNotificationsLongRunningTask singleton
-                wil::unique_cotaskmem_string unpackagedAppUserModelId;
-                RegisterUnpackagedApplicationHelper(GUID_NULL, unpackagedAppUserModelId); // create default registration for app
-
-                {
-                    auto lock = s_activatorInfoLock.lock_exclusive();
-                    s_protocolRegistration = true;
-                }
+                return first.Current().Value();
             }
+            ++first;
+        }
+        return nullptr;
+    }
 
-            BackgroundTaskBuilder builder{ nullptr };
+    void PushNotificationManager::Register()
+    {
+        winrt::hresult hr{ S_OK };
+        try
+        {
+            {
+                auto lock{ m_lock.lock_exclusive() };
+                THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_OPERATION_IN_PROGRESS), m_registering, "Registration is in progress!");
+                m_registering = true;
+            }
 
-            if (WI_IsFlagSet(registrationActivators, PushNotificationRegistrationActivators::PushTrigger) && PushNotificationHelpers::IsPackagedAppScenario())
+            auto scopeExitToCleanRegistrations{ wil::scope_exit([&]()
             {
-                GUID taskClsid = details.TaskClsid();
-                THROW_HR_IF(E_INVALIDARG, taskClsid == GUID_NULL);
+                auto lock { m_lock.lock_exclusive() };
+                m_registering = false;
+            }
+            ) };
 
+            if (PushNotificationHelpers::IsPackagedAppScenario())
+            {
                 {
-                    auto lock = s_activatorInfoLock.lock_shared();
-                    THROW_HR_IF(E_INVALIDARG, s_pushTriggerRegistration);
+                    auto lock{ m_lock.lock_shared() };
+                    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_REGISTERED), m_comActivatorRegistration || m_pushTriggerRegistration);
                 }
 
-                winrt::hstring taskClsidStr{ winrt::to_hstring(taskClsid) };
-                winrt::hstring backgroundTaskFullName{ backgroundTaskName + taskClsidStr };
-
-                auto tasks{ BackgroundTaskRegistration::AllTasks() };
-                bool isTaskRegistered{ std::any_of(std::begin(tasks), std::end(tasks),
-                    [&](auto&& task)
+                auto scopeExitToCleanRegistrations{ wil::scope_exit([&]()
```

Comment:
name collision with scope exit above. It will compile but please rename to avoid confusion.